### DISCUSSION TITLE
Possible to get "real" damage?

### DISCUSSION THREAD

**Discussion by xGuysOG** [saul/demofile-net#95] [Q&A]
I want to currently track how much damage a player has done in a round i made a small test projekt to compare what i can get from playerHurt event compared to what cs spits out, but playerhurt will spit out 100 damage even if the player had 1 health. Im unsure what is best pratice here, i considered at each round end one could save the matchstats object for the player for some of the stats.

and if one just wants to track on each side switch i guess one could pull matchstats on a side switch event, but i dont think we have one right now right?

@in0finite  Tagging you cuz you def know the answers to my dumb rambling, thank you for the help in advance.

Test code:
`    public static string flashWork = "C:\\Users\\xguys\\Desktop\\UnsignedFraggers_vs_ENABLE_map1_de_ancient_2024-09-16_18-00-20_66db4caddcd6bb3cd8b61a37_1726509620211\\66db4caddcd6bb3cd8b61a37_0_1726509620211.dem";

    public class ActualPlayerStats
    {
        public ulong steamID { get; set; }
        public int kills { get; set; }
        public int hsKills { get; set; }
        public int deaths { get; set; }
        public int assists { get; set; }
        public double homeMadeDamage { get; set; }
        public double DMG { get; set; }
    }
    public static Dictionary<ulong, ActualPlayerStats> actualPlayerStats = new Dictionary<ulong, ActualPlayerStats>();

    public static async Task Main(string[] args)
    {

        //return;
        try
        {
            Stopwatch stopwatch3 = new Stopwatch();
            stopwatch3.Start();
            MemoryStream _fileStream;
            byte[] _demoBytes;
            //_demoBytes = File.ReadAllBytes(demoFilePath);
            _demoBytes = await File.ReadAllBytesAsync(flashWork);
            _fileStream = new MemoryStream(_demoBytes);

            Stopwatch stopwatch = new Stopwatch();
            stopwatch.Start();
            var demo = new CsDemoParser();

            demo.Source1GameEvents.BombExploded += e =>
            {
                Console.WriteLine("Bomb exploded");
            };

            bool hasMatchEverStarted = false;
            demo.EntityEvents.CCSGameRulesProxy.AddChangeCallback(proxy => proxy.GameRules?.HasMatchStarted, (proxy, _, _) =>
            {
                hasMatchEverStarted |= demo.GameRules.HasMatchStarted;
            });

            demo.Source1GameEvents.PlayerHurt += e =>
            {
                if (!hasMatchEverStarted)
                {
                    return;
                }
                //if (e.Weapon != null && e.Weapon == "inferno")
                //{
                //    Console.WriteLine("Player hurt special called " + e.Player.PlayerName + " " + e.DmgHealth + " " + e.DmgArmor + " " + e.Weapon + " by " + e.Attacker.PlayerName);
                //    return;
                //}
                if (e.Attacker == null)
                {
                    return;
                }
                if (e.DmgHealth > 100)
                {
                    e.DmgHealth = 100;
                }
                Console.WriteLine("Player hurt called " + e.Attacker.PlayerName + " damagde with " + e.DmgHealth + " player Health " + e.Health  + " " + e.Weapon + " also current players health?" + e.Player.Health + " and pawn? " + e.PlayerPawn.Health + " and " + e.Player.PawnHealth);

                //make sure player is in actualPlayer stats or add it
                if (!actualPlayerStats.ContainsKey(e.Attacker.SteamID))
                {
                    actualPlayerStats.Add(e.Attacker.SteamID, new ActualPlayerStats()
                    {
                        steamID = e.Attacker.SteamID,
                        homeMadeDamage = e.DmgHealth,
                    });
                } else
                {
                    actualPlayerStats[e.Attacker.SteamID].homeMadeDamage += e.DmgHealth;
                }

            };

            demo.Source1GameEvents.RoundEnd += e =>
            {
                if (demo.GameRules.CSGamePhase != CSGamePhase.MatchEnded)
                    return;

                var table = new Table();

                table.AddColumn("Ping", column => column.Alignment = Justify.Right);
                table.AddColumn("Name");
                table.AddColumn("Money", column => column.Alignment = Justify.Right);
                table.AddColumn("Kills");
                table.AddColumn("Deaths");
                table.AddColumn("Assists");
                table.AddColumn("HS%");
                table.AddColumn("DMG");
                table.AddColumn("HomeMadeDMG");

                AddTeamRows(demo, table, demo.TeamCounterTerrorist);
                table.AddEmptyRow();
                AddTeamRows(demo, table, demo.TeamTerrorist);
                table.AddEmptyRow();
                AddTeamRows(demo, table, demo.TeamSpectator);

                AnsiConsole.Write(table);
            };

            demo.Source1GameEvents.CsWinPanelMatch += e =>
            {
                Console.WriteLine("WIn match has ended");
            };
            await demo.ReadAllAsync(_fileStream);
            stopwatch3.Stop();

            stopwatch.Stop();
            Console.WriteLine("Time elapsed after file data: {0}", stopwatch.Elapsed);
            Console.WriteLine("Time elapsed: {0}", stopwatch3.Elapsed);
            Console.WriteLine("\nFinished!");
        }
        catch (Exception e)
        {
            Console.WriteLine("Exception: " + e.Message.ToString());
        }

        Console.ReadKey();
    }

    private static void AddTeamRows(CsDemoParser demo, Table table, CCSTeam team)
    {
        var orderedPlayers = team.CSPlayerControllers
            .OrderByDescending(player => player.ActionTrackingServices!.MatchStats.Damage)
            .ToArray();

        foreach (var player in orderedPlayers)
        {
            var matchStats = player.ActionTrackingServices?.MatchStats ?? new CSMatchStats();
            var teamColour = player.CSTeamNum switch
            {
                CSTeamNumber.Terrorist => "bold red",
                CSTeamNumber.CounterTerrorist => "bold blue",
                _ => "grey"
            };

            // Render the player's name in their team colour, with a link to the SteamID
            var playerNameMarkup = new Markup(
                player.PlayerName,
                Style.Parse(teamColour)
                    .Link("https://steamcommunity.com/profiles/" + player.SteamID));

            if (player.CSTeamNum == CSTeamNumber.Spectator)
            {
                table.AddRow(new Markup("SPEC"), playerNameMarkup);
                continue;
            }

            table.AddRow(
                new Markup(player.Ping.ToString()),
                playerNameMarkup,
                new Markup("$" + (player.InGameMoneyServices?.Account ?? 0)),
                new Markup(matchStats.Kills.ToString()),
                new Markup(matchStats.Deaths.ToString()),
                new Markup(matchStats.Assists.ToString()),
                new Markup((matchStats.Kills > 0 ? matchStats.HeadShotKills * 100 / matchStats.Kills : 0).ToString()),
                new Markup(matchStats.Damage.ToString(), Style.Parse("bold")),
                new Markup(actualPlayerStats[player.SteamID].homeMadeDamage.ToString()));
        }
    }`

**Comment by in0finite** [comment]
Yes, accessing MatchStats or PerRoundStats is the correct approach.

As for teams switching sides, I am not aware of any event for this. But you can simply check the CSTeam of a player when a round ends.

**Comment by in0finite** [comment]
I just remembered that many properties inside `CSPerRoundStats` have `AllPlayersDuringFreezePeriodOrMatchEnd` attribute. This means that the server will only update them during FreezeTime or MatchEnd. So, if you access them during RoundEnd, they will have invalid values.

You will have to either access them during next RoundStart (FreezeTime), or when match ends.

**Reply by xGuysOG** [reply]
Thank you for the reply, i did some testing tho and im abit confused with the PerRoundStats. I did a simple test to see if the damage and other stats, lines up. i could just be completely misunderstand how it works, but yeah, dosnt corollate currently.
```
            demo.Source1GameEvents.RoundEnd += e =>
            {
                if (!hasMatchEverStarted)
                {
                    return;
                }

                var orderedPlayers = demo.TeamCounterTerrorist.CSPlayerControllers
.OrderByDescending(player => player.ActionTrackingServices!.MatchStats.Damage)
.ToArray();

                if (demo.GameRules.CSGamePhase != CSGamePhase.MatchEnded)
                    return;
                foreach (var player in orderedPlayers)
                {
                    if (player.SteamID != 76561198160006089)
                    {
                        continue;
                    }
                    var actionTracking = player.ActionTrackingServices;
                    if (actionTracking == null) continue;

                    var roundStats = actionTracking.PerRoundStats;
                    var matchStats = actionTracking.MatchStats;
                    int testDamage = 0;
                    int testKills = 0;
                    int testDeaths = 0;
                    // Loop through each round's stats
                    for (int roundNumber = 0; roundNumber < roundStats.Count; roundNumber++)
                    {
                        var stats = roundStats[roundNumber]; // Get the stats for the current round

                        // Print the player's name, steamID, damage, and round number
                        Console.WriteLine($"Player: {player.PlayerName} (SteamID: {player.SteamID})");
                        Console.WriteLine($"  - Round Number: {roundNumber}");
                        Console.WriteLine($"  - Damage: {stats.Damage}");
                        Console.WriteLine(); // Empty line for better readability
                        testDamage += stats.Damage;
                        testDeaths += stats.Deaths;
                        testKills += stats.Kills;
                    }
                    Console.WriteLine($"Total Test damage: {testDamage}");
                    Console.WriteLine($"Total damage: {matchStats.Damage}");
                    Console.WriteLine($"Total Test kills: {testKills}");
                    Console.WriteLine($"Total kills: {matchStats.Kills}");
                    Console.WriteLine($"Total Test Deaths: {testDeaths}");
                    Console.WriteLine($"Total Deaths: {matchStats.Deaths}");

                    Console.WriteLine($"Final score was " + demo.TeamCounterTerrorist.Score + " - " + demo.TeamTerrorist.Score);

                }
```
                
**and i got:**
  Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 0
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 1
  - Damage: 229

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 2
  - Damage: 205

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 3
  - Damage: 127

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 4
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 5
  - Damage: 33

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 6
  - Damage: 198

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 7
  - Damage: 63

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 8
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 9
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 10
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 11
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 12
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 13
  - Damage: 26

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 14
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 15
  - Damage: 5

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 16
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 17
  - Damage: 100

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 18
  - Damage: 137

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 19
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 20
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 21
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 22
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 23
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 24
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 25
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 26
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 27
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 28
  - Damage: 0

Player: D1DRIK (SteamID: 76561198160006089)
  - Round Number: 29
  - Damage: 0

Total Test damage: 1123
Total damage: 1445
Total Test kills: 10
Total kills: 13
Total Test Deaths: 11
Total Deaths: 14
Final score was 13 - 6

**Comment by in0finite** [comment]
I tested `MatchStats` against HLTV statistics, and it matches. However, you're right, `PerRoundStats` is not identical, for some reason. So I would subscribe to `RoundStart` event (because that's the time when properties are updated), and calculate last round damage as difference between current and last `MatchStats.Damage`.
